src/request.js

```diff
@@ -0,0 +1,237 @@
+
+  assert(http.METHODS.indexOf(method) !== -1);
+  
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.data, 'if supplying a stream, you must not supply data');
+  }
+
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.stream, 'if supplying a stream, you must not supply data');
+  }
+
+  switch (urlParts.protocol) {
+    case 'https:':
+      return makeRequest({
+        httpLib: https,
+        hostname: urlParts.hostname,
+        port: port || 443,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    case 'http:':
+      return makeRequest({
+        httpLib: http,
+        hostname: urlParts.hostname,
+        port: port || 80,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    default:
+      throw new RequestError('Invalid protocol: ' + urlParts.protocol);
+      break;
+  }
+  
+}
+
+module.exports = {
+  request,
+  RequestError,
+};
+
+async function makeRequest(opts) {
+  assert(typeof opts === 'object');
+  assert(typeof opts.httpLib === 'object');
+  assert(typeof opts.hostname === 'string');
+  assert(typeof opts.port === 'number');
+  assert(typeof opts.path === 'string');
+  assert(typeof opts.method === 'string');
+  assert(typeof opts.headers === 'object');
+  assert(typeof opts.timeout === 'number');
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.stream === 'object');
+    assert(typeof opts.stream.pipe === 'function');
+  }
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.data.length === 'number');
+  }
+
+  return new Promise(async (resolve, reject) => {
+    let request = opts.httpLib.request({
+      hostname: opts.hostname,
+      port: opts.port,
+      path: opts.path,
+      method: opts.method,
+    });
+
+    // We don't want connections hanging around forever, so we'll set a fatal
+    // condition that they timeout if they aren't responded to quickly enough
+    request.setTimeout(opts.timeout);
+
+    request.on('error', err => {
+      reject(err);
+    });
+
+    request.on('socket', socket => {
+      debug('have a socket');
+
+      socket.on('error', err => {
+        debug('socket error');
+        reject(err);
+      });
+
+      socket.on('close', had_error => {
+        reject(new RequestError('Socket closed with error'));
+      });
+
+    });
+
+    request.on('response', response => {
```

Comment:
> or do you know that they bubble up into being http/https errors?

If socket errors don't bubble up that would be a major bug in node/http.
I'm pretty sure they they do as I've never listened for socket errors. But triggering them caused things higher up.

if you must know check the source:
https://github.com/nodejs/node/blob/master/lib/_http_client.js#L559
https://github.com/nodejs/node/blob/master/lib/_http_client.js#L309

But docs are pretty clear on this... Otherwise a socket error would crash the process, if the http library didn't handle them and propagate things.